### PR TITLE
Feature/39702 school pin gen fallback

### PR DESCRIPTION
**Pull request by GuyHarwood**
In case of school pin timed function failing, this is a fallback that calls the pin gen function via http at the point of pupil pin generation to ensure a school password is generated.

Tasks
------
- [x] remove  sp_addmessage migrations
- [x] add feature toggle for school pin fallback http call

### CODE REVIEW

File: admin/data/sql/migrations/20200302135508.do.spcreatechecks-add-school-pin-expiry.sql [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+    -- Use RAISERROR inside the CATCH block to return
+    -- error information about the original error that
+    -- caused execution to jump to the CATCH block.
+    RAISERROR (@ErrorMessage, -- Message text.
+        @ErrorSeverity, -- Severity.
+        @ErrorState -- State.
+        );
+END CATCH
+GO
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200302135508.do.spcreatechecks-add-school-pin-expiry.sql]: The `GO` is not required.

File: admin/data/sql/migrations/20200302135508.undo.spcreatechecks-add-school-pin-expiry.sql [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    -- Use RAISERROR inside the CATCH block to return
+    -- error information about the original error that
+    -- caused execution to jump to the CATCH block.
+    RAISERROR (@ErrorMessage, -- Message text.
+        @ErrorSeverity, -- Severity.
+        @ErrorState -- State.
+        );
+END CATCH
+GO
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20200302135508.undo.spcreatechecks-add-school-pin-expiry.sql]: `GO` not required

File: load-test/bin/undo-generate-pupil-load-test-data.sql [hunk 3]
```diff
@@ -12,12 +12,11 @@ truncate table mtc_admin.checkConfig
 -- disable the next line for data volume testing
 DELETE FROM mtc_admin.[check]
 truncate table mtc_admin.auditLog
+
 truncate table mtc_admin.pupilAccessArrangements
+truncate table mtc_admin.pupilAttendance
```
Comment (jon-shipley) [inline load-test/bin/undo-generate-pupil-load-test-data.sql]: Should be consistently upper or lowercase.

File: tslib/src/functions/school-pin-http-service/index.ts [hunk 4]
```diff
@@ -13,19 +13,27 @@ function finish (start: number, context: Context) {
...
   }
 
+  const schoolId = req.body.school_id
+
   const start = performance.now()
   const schoolPinReplenishmentService = new SchoolPinReplenishmnentService()
-  await schoolPinReplenishmentService.process(context.log, schoolUuid)
+  context.log(`requesting pin for school:${schoolId}`)
+  const newPin = await schoolPinReplenishmentService.process(context.log, schoolId)
+  context.log(`pin created is ${newPin}`)
```
Comment (jon-shipley) [inline tslib/src/functions/school-pin-http-service/index.ts]: IMO This log needs a school identifier to be meaningful in production.

**Comment by GuyHarwood** [comment]
calls to `sp_addMessage` are not compatible with sql azure.  taking back into development for refactor.

File: admin/services/check-start.service/check-start.service.spec.js [hunk 5]
```diff
@@ -134,40 +136,125 @@ describe('check-start.service', () => {
...
+          1: configService.getBaseConfig(),
+          2: configService.getBaseConfig(),
+          3: configService.getBaseConfig()
+        })
+      spyOn(checkStartDataService, 'sqlStoreBatchConfigs')
+      spyOn(redisCacheService, 'get').and.returnValue(Promise.resolve())
+      spyOn(redisCacheService, 'set').and.returnValue(Promise.resolve())
+    })
+
+    test('attempts to generate school pin if error 50001 thrown and setting enabled', async () => {
```
Comment (GeorgeChatzigiannis) [inline admin/services/check-start.service/check-start.service.spec.js]: `is enabled` ? 🤔

File: admin/services/check-start.service/check-start.service.spec.js [hunk 6]
```diff
@@ -134,40 +136,125 @@ describe('check-start.service', () => {
...
+        await checkStartService.prepareCheck2(pupilIds, dfeNumber, schoolId, true, null, checkWindowMock)
+        fail('error should have been thrown')
+      } catch (error) {
+        expect(error).toBeDefined()
+        expect(schoolPinService.generateSchoolPin).toHaveBeenCalledTimes(1)
+        expect(pinGenerationDataService.sqlCreateBatch).toHaveBeenCalledTimes(2)
+      }
+    })
+
+    test('does not attempt to generate school pin if error 50001 thrown and setting disabled', async () => {
```
Comment (GeorgeChatzigiannis) [inline admin/services/check-start.service/check-start.service.spec.js]: is disabled?

File: admin/services/check-start.service/check-start.service.spec.js (hunk 5 above)
Comment (GuyHarwood) [reply admin/services/check-start.service/check-start.service.spec.js]: both are valid english i think

File: admin/services/check-start.service/check-start.service.spec.js (hunk 6 above)
Comment (GuyHarwood) [reply admin/services/check-start.service/check-start.service.spec.js]: both are valid english i think

**Pull request merged**